File: app/src/main/java/com/boringdroid/systemui/AppLoaderTask.java
package com.boringdroid.systemui;

import android.content.Context;
import android.content.pm.LauncherActivityInfo;
import android.content.pm.LauncherApps;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.UserHandle;
import android.os.UserManager;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

public class AppLoaderTask implements Runnable {
    private static final HandlerThread WORK_THREAD = new HandlerThread("app-loader-thread");

    static {
        WORK_THREAD.start();
    }

    private final Handler sHandler = new Handler(WORK_THREAD.getLooper());

    private final WeakReference<Context> mContext;
    private final WeakReference<Handler> mTarget;

    private final List<AppData> mAllApps = new ArrayList<>();
    private boolean mStopped = false;

    public AppLoaderTask(Context context, Handler target) {
        mContext = new WeakReference<>(context);
        mTarget = new WeakReference<>(target);
    }

    @Override
    public void run() {
        if (mStopped) {
            return;
        }
        Context context = getContext();
        if (context == null) {
            return;
        }
        UserManager userManager = (UserManager) context.getSystemService(Context.USER_SERVICE);
        List<UserHandle> userHandles = userManager.getUserProfiles();
        List<LauncherActivityInfo> activityInfoList = new ArrayList<>();
        LauncherApps launcherApps =
                (LauncherApps) context.getSystemService(Context.LAUNCHER_APPS_SERVICE);
        for (UserHandle userHandle : userHandles) {
            activityInfoList.addAll(launcherApps.getActivityList(null, userHandle));
        }
        mAllApps.clear();
        for (LauncherActivityInfo info : activityInfoList) {
            AppData appInfo = new AppData();
            appInfo.setName((String) info.getLabel());
            appInfo.setComponentName(info.getComponentName());
            appInfo.setPackageName(info.getApplicationInfo().packageName);
            appInfo.setIcon(info.getIcon(0));
            mAllApps.add(appInfo);
        }
        mAllApps.sort(
                (appDataOne, appDataTwo) -> appDataOne.getName().compareTo(appDataTwo.getName()));
        Handler target = getTarget();
        if (target != null) {
            target.sendEmptyMessage(HandlerConstant.H_LOAD_SUCCEED);
        }
    }

    public synchronized void start() {
        mStopped = false;
        sHandler.post(this);
    }

    public synchronized void stop() {
        mStopped = true;
        notify();
    }

    public List<AppData> getAllApps() {
        return new ArrayList<>(mAllApps);
    }

    private Handler getTarget() {
        return mTarget.get();
    }

    private Context getContext() {
        return mContext.get();
    }
}

File: app/src/main/java/com/boringdroid/systemui/AppLoaderTask.kt
package com.boringdroid.systemui

import android.content.Context
import android.content.pm.LauncherActivityInfo
import android.content.pm.LauncherApps
import android.os.Handler
import android.os.HandlerThread
import android.os.UserManager
import java.lang.ref.WeakReference

class AppLoaderTask(context: Context?, target: Handler?) : Runnable {
    companion object {
        private val WORK_THREAD = HandlerThread("app-loader-thread")

        init {
            WORK_THREAD.start()
        }
    }

    private val handler = Handler(WORK_THREAD.looper)
    private val loaderContext: WeakReference<Context?>?
    private val loaderTarget: WeakReference<Handler?>?
    private val loaderAllApps: MutableList<AppData> = ArrayList()
    private var stopped = false
    override fun run() {
        if (stopped) {
            return
        }
        val context = context ?: return
        val userManager = context.getSystemService(Context.USER_SERVICE) as UserManager
        val userHandles = userManager.userProfiles
        val activityInfoList: MutableList<LauncherActivityInfo> = ArrayList()
        val launcherApps = context.getSystemService(Context.LAUNCHER_APPS_SERVICE) as LauncherApps
        for (userHandle in userHandles) {
            activityInfoList.addAll(launcherApps.getActivityList(null, userHandle))
        }
        loaderAllApps.clear()
        for (info in activityInfoList) {
            val appData = AppData()
            appData.name = info.label as String
            appData.componentName = info.componentName
            appData.packageName = info.applicationInfo.packageName
            appData.icon = info.getIcon(0)
            loaderAllApps.add(appData)
        }
        loaderAllApps.sortWith { appDataOne: AppData, appDataTwo: AppData ->
            appDataOne.name!!.compareTo(
                appDataTwo.name!!
            )
        }
        val target = target
        target?.sendEmptyMessage(HandlerConstant.H_LOAD_SUCCEED)
    }

    @Synchronized
    fun start() {
        stopped = false
        handler.post(this)
    }

    @Synchronized
    fun stop() {
        stopped = true
        // Could we remove notify() from kotlin
        // notify()
    }

    val allApps: List<AppData>
        get() = ArrayList(loaderAllApps)
    private val target: Handler?
        get() = if (loaderTarget?.get() != null) loaderTarget.get() else null
    private val context: Context?
        get() = if (loaderContext?.get() != null) loaderContext.get() else null

    init {
        loaderContext = WeakReference(context)
        loaderTarget = WeakReference(target)
    }
}
